src/core/lib/iomgr/parse_address.cc

```diff
@@ -56,27 +56,44 @@ bool grpc_parse_unix(const grpc_uri* uri,
   if (path_len == maxlen) return false;
   un->sun_family = AF_UNIX;
   strcpy(un->sun_path, uri->path);
-  if (un->sun_path[0] == '@') {
-    // The address length of an abstract unix socket is calculated differently
-    // from pathname sockets. Effectively what it means for us is that we *must
-    // not* take a string terminating null byte into account. See `man 7 unix`
-    // for more details.
-    resolved_addr->len =
-        static_cast<socklen_t>(sizeof(un->sun_family) + path_len);
-    un->sun_path[0] = '\0';
-    return true;
-  }
   resolved_addr->len = static_cast<socklen_t>(sizeof(*un));
   return true;
 }
 
+bool grpc_parse_unix_abstract(const grpc_uri* uri,
+                              grpc_resolved_address* resolved_addr) {
+  if (strcmp("unix-abstract", uri->scheme) != 0) {
+    gpr_log(GPR_ERROR, "Expected 'unix-abstract' scheme, got '%s'", uri->scheme);
+    return false;    
+  }
+  struct sockaddr_un* un =
+      reinterpret_cast<struct sockaddr_un*>(resolved_addr->addr);
+  const size_t maxlen = sizeof(un->sun_path);
+  const size_t path_len = strnlen(uri->path, maxlen);
+  if (path_len == maxlen) return false;
+  un->sun_family = AF_UNIX;
+
+  // Abstract sockets paths begin with a null charcter.
+  un->sun_path[0] = '\0';
+  // The address length of an abstract unix socket does not include a terminating null.
+  strcpy(un->sun_path + 1, uri->path);
```

Comment:
I don't think this will work correctly in the presence of NUL bytes within the string.  We may actually need to change the URI parser code to use `std::string` instead of C-style `char*`, since the latter probably won't work for this in the first place.

We should also test this case to make sure it works properly.

Reply:
Do we want/need to support null bytes within the string? That's a very not-c-friendly thing to do.

Reply:
As we discussed offline, abstract sockets do allow NUL bytes in the name, so we should support them here.  But since this requires changes to the URI parser, I'm okay with fixing that separately.